Fixes 422 error due to bad sql building

cyrillefr opened the pull request:
#### What? Why?

- Closes #12656

When some properties are defined at the enterprise level, there is a 422 return code as well as a "Products loading ..." displayed when customer filters products.

Reason is a bug in SQL Query building.
First part of query use `supplier_properties` parameter, but not second part, that can leads to mismatch between the 2 parts:
https://github.com/openfoodfoundation/openfoodnetwork/blob/dfea0cd805c555b84dce163f9ed83842a5b5b992/app/services/order_cycles/distributed_products_service.rb#L37-L40

The 2nd bit:
https://github.com/openfoodfoundation/openfoodnetwork/blob/dfea0cd805c555b84dce163f9ed83842a5b5b992/app/services/order_cycles/distributed_products_service.rb#L81-L83


#### What should we test?
See § "Steps to Reproduce" from #12656

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [X] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

cyrillefr commented:
Hello @rioug ,

I uploaded a bug here in the `order` method (the `||` is not good).

I have working on the `order` method locally, it looks a bit like that

```ruby
elsif sorting_with_supplier_properties?(supplier_properties)
  order_by_property = distributor.producer_properties 
    .map { |pp| "producer_properties.property_id=#{pp.property_id} DESC" } 
    .join(", ")
     "#{order_by_property}, spree_products.name ASC, spree_products.id ASC"
```
```ruby
def sorting_with_supplier_properties?(supplier_properties)
  supplier_properties && distributor.producer_properties.present?
end
```

I am not sure whether my use of `distributor.producer_properties` is good( I have seen it used in specs).
But I realise now that may be I should make another `if` in the `relation_by_sorting` method, to use
`producer_properties.property_id` in the `select` and in the `group by`.

What do you think ?

rioug commented:
If I am following correctly what you are saying, I think you are trying to something we don't need. There is no need to order by `producer_properties`.

I think what we need to do here, it's to get rid of the `supplier_properties` flag (and the `||` ) in `relation_by_sorting` and update the the query when we are sorting "by_category", we need to include the `first_variant.supplier_id` in the select. Doing that allows us to use `supplier_property_join` when sorting "by_category" without blowing up. Does that make sense ? I have a feeling that might not work but it's a good place to start.

Good job on cleaning up  `order !`

rioug commented:
As a first step you should try to add the missing spec in here : 
https://github.com/openfoodfoundation/openfoodnetwork/blob/dfea0cd805c555b84dce163f9ed83842a5b5b992/spec/services/products_renderer_spec.rb#L139
We need the scenario where `preferred_shopfront_taxon_order` is set on the distributor and we are filtering by producer properties.

cyrillefr commented:
Hello @rioug ,

Thanks for the tip.
It seems I handled this issue from the wrong end with this supplier_properties parameter ....
So, I removed it from the `relation_by_sorting`.
I added `supplier_id` in the 2nd part of the method, and it works.
I also added one spec.

Checks have failed, but I cannot reproduce one and the 2nd fail is a redis issue ...
Could you please rerun the checks ? It should be ok

rioug reviewed (approved):
Awesome, thanks for fixing this @cyrillefr :pray: ! I had a feeling it wasn't going to be this simple but I am glad to be proven wrong :smile:

dacook commented on spec/services/products_renderer_spec.rb:
```diff
@@ -155,6 +155,19 @@
+          search_param = { q: { "with_variants_supplier_properties" => [property_organic.id] } }
+          products_renderer = ProductsRenderer.new(distributor, order_cycle, customer, search_param)
+
+          products = products_renderer.send(:products)
```
According to our rubocop rules, this needs to use the `__send__` method.
Rubocop was skipping this file, but we recently enabled it (after this branch was made).
I will go ahead and correct it.
```suggestion
          products = products_renderer.__send__(:products)
```

dacook replied:
It's worth noting that usage of `send`/`__send__` is generally not good practice, and we should be testing the result of the public methods. So next time, we should try to do that. This is fine for now though.

filipefurtad0 commented:
Thanks so much @cyrillefr ,

I could verify the bug is fixed indeed in staging. Before this PR, the bug is reproducible:

![image](https://github.com/user-attachments/assets/d19562a6-bec9-4306-9ad0-67afa67cc05a)


After this PR the bug is gone, and the correct results are returned:
![image](https://github.com/user-attachments/assets/063f70b2-2ade-4fa3-bb6f-79594497ae24)

Merging!
Thank you.
